File: benchmarks/JavaScript/benchmark.js
'use strict';

function Benchmark() {
  var that = this;

  this.innerBenchmarkLoop = function (innerIterations) {
    for (var i = 0; i < innerIterations; i++) {
      if (!that.verifyResult(that.benchmark())) {
        return false;
      }
    }
    return true;
  };

  this.benchmark = function () {
    throw "subclass responsibility"
  };

  this.verifyResult = function () {
    throw "subclass responsibility"
  };
}

exports.Benchmark = Benchmark;

File: benchmarks/Rust/src/benchmark.rs
use std::any::Any;

pub trait Benchmark {
    fn benchmark(&self) -> Box<dyn Any>;
    fn verify_result(&self, result: Box<dyn Any>) -> bool;

    fn inner_benchmark_loop(&self, inner_iterations: usize) -> bool {
        for _ in 0..inner_iterations {
            if !self.verify_result(self.benchmark()) {
                return false;
            }
        }
        true
    }
}
